xlators/cluster/dht/src/dht-inode-write.c

```diff
@@ -10,96 +10,76 @@
-    }
-
-    /* writev fails with EBADF if dht has not yet opened the fd
-     * on the cached subvol. This could happen if the file was migrated
-     * and a lookup updated the cached subvol in the inode ctx.
-     * We only check once as this could be a valid bad fd error.
-     */
+dht_inode_write2(xlator_t *this, xlator_t *subvol, call_frame_t *frame,
+                 int ret);
 
-    if (dht_check_remote_fd_failed_error(local, op_ret, op_errno)) {
-        ret = dht_check_and_open_fd_on_subvol(this, frame);
-        if (ret)
-            goto out;
-        return 0;
-    }
-
-    if (op_ret == -1 && !dht_inode_missing(op_errno)) {
-        local->op_errno = op_errno;
-        local->op_ret = -1;
-        gf_msg_debug(this->name, 0, "subvolume %s returned -1 (%s)", prev->name,
-                     strerror(op_errno));
-        goto out;
-    }
-
-    if (local->call_cnt != 1) {
-        /* preserve the modes of source */
-        if (local->stbuf.ia_blocks) {
-            dht_iatt_merge(this, postbuf, &local->stbuf);
-            dht_iatt_merge(this, prebuf, &local->prebuf);
-        }
-        goto out;
-    }
-
-    local->rebalance.target_op_fn = dht_writev2;
-
-    local->op_ret = op_ret;
-    local->op_errno = op_errno;
-
-    /* We might need to pass the stbuf information to the higher DHT
-     * layer for appropriate handling.
-     */
-
-    dht_set_local_rebalance(this, local, NULL, prebuf, postbuf, xdata);
-
-    /* Phase 2 of migration */
-    if ((op_ret == -1) || IS_DHT_MIGRATION_PHASE2(postbuf)) {
-        ret = dht_rebalance_complete_check(this, frame);
-        if (!ret)
-            return 0;
+void
+dht_inode_write_unwind(glusterfs_fop_t fop, call_frame_t *frame, int32_t op_ret,
+                       int32_t op_errno, struct iatt *prestat,
+                       struct iatt *poststat, dict_t *xdata)
+{
+    switch (fop) {
+        case GF_FOP_WRITE:
+            DHT_STACK_UNWIND(writev, frame, op_ret, op_errno, prestat, poststat,
+                             xdata);
+            break;
+
+        case GF_FOP_ZEROFILL:
+            DHT_STACK_UNWIND(zerofill, frame, op_ret, op_errno, prestat,
+                             poststat, xdata);
+            break;
+
+        case GF_FOP_DISCARD:
+            DHT_STACK_UNWIND(discard, frame, op_ret, op_errno, prestat,
+                             poststat, xdata);
+            break;
+
+        case GF_FOP_TRUNCATE:
+            DHT_STACK_UNWIND(truncate, frame, op_ret, op_errno, prestat,
+                             poststat, xdata);
+            break;
+
+        case GF_FOP_FTRUNCATE:
+            DHT_STACK_UNWIND(ftruncate, frame, op_ret, op_errno, prestat,
+                             poststat, xdata);
+            break;
+
+        case GF_FOP_FALLOCATE:
+            DHT_STACK_UNWIND(fallocate, frame, op_ret, op_errno, prestat,
+                             poststat, xdata);
+            break;
+            /*TODO: too many cbk for dirs is different, cbk for mds is
```

Comment:
nit: unsure what this TODO means...

Reply:
Thanks, should have removed this. I fixed this one already in my patch. When I was refactoring I saw that for setattr fops there are multiple cbks. Only one patch could be refactored.

Reply:
I will remove this in the next version